include/circt/Dialect/FIRRTL/FIRRTLTypes.h

```diff
@@ -344,5 +353,70 @@ struct DenseMapInfo<circt::firrtl::FIRRTLType> {
 };
 
 } // namespace llvm
+namespace circt {
+namespace firrtl {
+template <typename... BaseTy>
+bool type_isa(Type type) { // NOLINT(readability-identifier-naming)
+  // First check if the type is the requested type.
+  if (type.isa<BaseTy...>())
+    return true;
+
+  // Then check if it is a type alias wrapping the requested type.
+  if (auto alias = type.dyn_cast<BaseTypeAliasType>())
+    return alias.getInnerType().isa<BaseTy...>();
+
+  return false;
+}
+
+// type_isa for a nullable argument.
+template <typename... BaseTy>
+bool type_isa_and_nonnull(Type type) { // NOLINT(readability-identifier-naming)
+  if (!type)
+    return false;
+  return type_isa<BaseTy...>(type);
+}
+
+template <typename BaseTy>
+BaseTy type_cast(Type type) { // NOLINT(readability-identifier-naming)
+  assert(type_isa<BaseTy>(type) && "type must convert to requested type");
+
+  // If the type is the requested type, return it.
+  if (type.isa<BaseTy>())
+    return type.cast<BaseTy>();
+
+  // Otherwise, it must be a type alias wrapping the requested type.
+  return type.cast<BaseTypeAliasType>().getInnerType().cast<BaseTy>();
+}
+
+template <typename BaseTy>
+BaseTy type_dyn_cast(Type type) { // NOLINT(readability-identifier-naming)
+  if (!type_isa<BaseTy>(type))
+    return BaseTy();
+
+  return type_cast<BaseTy>(type);
+}
+
+template <typename BaseTy>
+class BaseTypeAliasOr
+    : public ::mlir::Type::TypeBase<BaseTypeAliasOr<BaseTy>,
+                                    firrtl::FIRRTLBaseType,
```

Comment:
Bit wild but maayyybbbee this should inherit from `BaseTy` instead of `FIRRTLBaseType`?

Just thinking this would avoid the need for a LOT of `.get()` (previously, casting) by exposing all those methods.  Might that work? :crossed_fingers:

Reply:
Nevermind, would be bonkers re:type "storage" unfortunately but dang that'd be nice.  Or maybe there's a way :smile: .